File: proposals/wrt-004-model/input_manifest.py
#!/usr/bin/env python3
"""`warrant.verify-report@v1` sealed observation — Python reference.

Round 2 of WRT-004 §6, rebuilt after round 1 was **refuted**. Round 1 walked
the filesystem and hashed what it found; a live verifier follows a symlinked
record that the walk silently skipped, so `input_root` committed to `[]`
while the report judged one record. A manifest that does not commit the bytes
the judgement used is not a manifest.

**One atomic observation.** `seal()` produces the store's byte view exactly
once. Both `input_manifest` and — in a full implementation — the judgement
are derived from *that view and nothing else*, so they cannot disagree about
what exists. Anything the seal refuses is not judged either.

    input_manifest.py <store-dir> [--trust-config PATH]   # JCS bytes
    input_manifest.py <store-dir> --root                  # input_root only
"""

import hashlib
import json
import os
import sys

DOMAIN = b"warrant.verify-report.input@v1:"

# SPEC §4 short escapes. Everything else below U+0020 is \u00xx LOWERCASE;
# everything else — including < > & / and all non-ASCII, and explicitly
# U+2028/U+2029 — is raw UTF-8. Round 1's Go used encoding/json, which
# escapes U+2028 even with SetEscapeHTML(false), and that is forbidden here.
SHORT = {0x22: '\\"', 0x5C: "\\\\", 0x08: "\\b", 0x09: "\\t",
         0x0A: "\\n", 0x0C: "\\f", 0x0D: "\\r"}


def jcs_string(s):
    out = ['"']
    for ch in s:
        cp = ord(ch)
        if cp in SHORT:
            out.append(SHORT[cp])
        elif cp < 0x20:
            out.append("\\u%04x" % cp)
        else:
            out.append(ch)
    out.append('"')
    return "".join(out)


def jcs(value):
    """The SPEC §4 subset, written out rather than delegated to a library.

    `json.dumps(ensure_ascii=False)` happens to agree with §4 today, but the
    agreement is incidental — a library default is not a specification, and
    the Go side proved that by escaping U+2028 under the same intent.
    """
    if isinstance(value, str):
        return jcs_string(value)
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, int):
        return str(value)
    if value is None:
        return "null"
    if isinstance(value, list):
        return "[" + ",".join(jcs(v) for v in value) + "]"
    if isinstance(value, dict):
        # UTF-16 code-unit order, which is JCS's rule for keys. All keys here
        # are ASCII, where it coincides with code-point order; the sort key is
        # written out anyway so a future non-ASCII key does not silently
        # change meaning.
        items = sorted(value.items(), key=lambda kv: kv[0].encode("utf-16-be"))
        return "{" + ",".join(jcs_string(k) + ":" + jcs(v)
                              for k, v in items) + "}"
    raise TypeError("not representable in the §4 subset: %r" % type(value))


def role_of(rel):
    parts = rel.split("/")
    if parts[0] == "records" and len(parts) == 2 and rel.endswith(".json"):
        return "record"
    if parts[0] == "blobs" and len(parts) == 2:
        return "blob"
    if rel == "genesis.json":
        return "genesis"
    return "other"


def seal(store_dir, trust_config=None):
    """The observation. Every path the verifier can see, with its state.

    `state` is a sum type, because round 1's schema could not represent what
    §3.1 required: it demanded that failed reads appear while making
    `sha256` mandatory, and there are no bytes after a failed read.

      read       — bytes obtained; `sha256` present
      unreadable — attempted, no bytes (permissions, I/O)
      refused    — deliberately not read, with a reason

    A symlink is **refused**, not followed and not skipped. Following it
    leaves the store's byte universe; skipping it silently is what made the
    round-1 manifest disagree with the judgement. Refusing it is visible in
    the manifest, and the judgement — reading only this view — does not see
    the record either. That is a deliberate difference from `@v0`, which
    follows symlinks, and it is why `@v1` is a different tag.
    """
    view = []
    base = os.path.abspath(store_dir)
    for root, dirs, names in os.walk(base):
        dirs.sort()
        for name in sorted(names + [d for d in dirs
                                    if os.path.islink(os.path.join(root, d))]):
            full = os.path.join(root, name)
            rel = os.path.relpath(full, base).replace(os.sep, "/")
            e = {"path": rel, "role": role_of(rel)}
            if os.path.islink(full):
                e["state"], e["reason"] = "refused", "symlink"
            elif not os.path.isfile(full):
                e["state"], e["reason"] = "refused", "not-a-regular-file"
            else:
                try:
                    with open(full, "rb") as fh:
                        raw = fh.read()
                except OSError:
                    e["state"] = "unreadable"
                else:
                    e["state"] = "read"
                    e["sha256"] = hashlib.sha256(raw).hexdigest()
            view.append(e)
    if trust_config:
        rel = os.path.basename(trust_config)
        e = {"path": rel, "role": "trust-config"}
        try:
            with open(trust_config, "rb") as fh:
                raw = fh.read()
        except OSError:
            e["state"] = "unreadable"
        else:
            e["state"] = "read"
            e["sha256"] = hashlib.sha256(raw).hexdigest()
        view.append(e)
    view.sort(key=lambda e: e["path"].encode("utf-8"))
    paths = [e["path"] for e in view]
    if len(set(paths)) != len(paths):
        raise ValueError("duplicate path in the observation: refused")
    return view


def input_root(view):
    return hashlib.sha256(DOMAIN + jcs(view).encode("utf-8")).hexdigest()


def main(argv):
    if len(argv) < 2:
        print(__doc__.strip().splitlines()[-2])
        return 2
    trust = argv[argv.index("--trust-config") + 1] \
        if "--trust-config" in argv else None
    try:
        view = seal(argv[1], trust)
    except ValueError as exc:
        print("REFUSED: %s" % exc, file=sys.stderr)
        return 1
    if "--root" in argv:
        sys.stdout.write(input_root(view) + "\n")
    else:
        sys.stdout.buffer.write(jcs(view).encode("utf-8"))
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv))

File: proposals/wrt-004-model/input_manifest.go
// `warrant.verify-report@v1` sealed observation — Go reference.
//
// Round 2 of WRT-004 §6, rebuilt after round 1 was refuted. Two defects were
// mine, in this file: the walk silently skipped symlinks a live verifier
// follows, and `encoding/json` escaped U+2028 even with SetEscapeHTML(false),
// which SPEC §4 forbids outright. The encoder below is written out rather
// than delegated, for that reason.
//
// One atomic observation: seal() produces the byte view once, and everything
// else derives from it, so the manifest and the judgement cannot disagree
// about what exists.
//
//	input_manifest <store-dir> [--trust-config PATH]   # JCS bytes
//	input_manifest <store-dir> --root                  # input_root only
package main

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"unicode/utf16"
)

const domain = "warrant.verify-report.input@v1:"

type entry struct {
	Path   string
	Role   string
	State  string // read | unreadable | refused
	Sha256 string // present iff State == "read"
	Reason string // present iff State == "refused"
}

// jcsString implements SPEC §4 string escaping exactly: the seven short
// escapes, \u00xx LOWERCASE below U+0020, and everything else raw UTF-8 —
// including < > & / and U+2028/U+2029, which encoding/json escapes and §4
// says MUST NOT be escaped.
func jcsString(s string) string {
	var b strings.Builder
	b.WriteByte('"')
	for _, r := range s {
		switch r {
		case '"':
			b.WriteString(`\"`)
		case '\\':
			b.WriteString(`\\`)
		case '\b':
			b.WriteString(`\b`)
		case '\t':
			b.WriteString(`\t`)
		case '\n':
			b.WriteString(`\n`)
		case '\f':
			b.WriteString(`\f`)
		case '\r':
			b.WriteString(`\r`)
		default:
			if r < 0x20 {
				b.WriteString(fmt.Sprintf(`\u%04x`, r))
			} else {
				b.WriteRune(r)
			}
		}
	}
	b.WriteByte('"')
	return b.String()
}

// utf16Less orders keys by UTF-16 code units, which is JCS's rule. Every key
// here is ASCII, where it coincides with byte order; it is written out so a
// future non-ASCII key does not silently change meaning.
func utf16Less(a, b string) bool {
	ua, ub := utf16.Encode([]rune(a)), utf16.Encode([]rune(b))
	for i := 0; i < len(ua) && i < len(ub); i++ {
		if ua[i] != ub[i] {
			return ua[i] < ub[i]
		}
	}
	return len(ua) < len(ub)
}

// jcsEntry emits one entry as a JCS object with its members in key order.
// The optional members are omitted, not emitted as null: an entry that was
// never read has no digest, and "sha256": null would be a claim about bytes
// that do not exist.
func jcsEntry(e entry) string {
	kv := map[string]string{
		"path": jcsString(e.Path), "role": jcsString(e.Role),
		"state": jcsString(e.State),
	}
	if e.State == "read" {
		kv["sha256"] = jcsString(e.Sha256)
	}
	if e.State == "refused" {
		kv["reason"] = jcsString(e.Reason)
	}
	keys := make([]string, 0, len(kv))
	for k := range kv {
		keys = append(keys, k)
	}
	sort.Slice(keys, func(i, j int) bool { return utf16Less(keys[i], keys[j]) })
	parts := make([]string, 0, len(keys))
	for _, k := range keys {
		parts = append(parts, jcsString(k)+":"+kv[k])
	}
	return "{" + strings.Join(parts, ",") + "}"
}

func jcsView(view []entry) string {
	parts := make([]string, 0, len(view))
	for _, e := range view {
		parts = append(parts, jcsEntry(e))
	}
	return "[" + strings.Join(parts, ",") + "]"
}

func roleOf(rel string) string {
	parts := strings.Split(rel, "/")
	if parts[0] == "records" && len(parts) == 2 && strings.HasSuffix(rel, ".json") {
		return "record"
	}
	if parts[0] == "blobs" && len(parts) == 2 {
		return "blob"
	}
	if rel == "genesis.json" {
		return "genesis"
	}
	return "other"
}

// seal is the observation. A symlink is REFUSED — not followed, not skipped:
// following it leaves the store's byte universe, and skipping it silently is
// what made round 1's manifest disagree with the judgement.
func seal(storeDir, trustConfig string) ([]entry, error) {
	base, err := filepath.Abs(storeDir)
	if err != nil {
		return nil, err
	}
	var view []entry
	err = filepath.WalkDir(base, func(p string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if p == base {
			return nil
		}
		rel, err := filepath.Rel(base, p)
		if err != nil {
			return err
		}
		rel = filepath.ToSlash(rel)
		e := entry{Path: rel, Role: roleOf(rel)}
		switch {
		case d.Type()&fs.ModeSymlink != 0:
			e.State, e.Reason = "refused", "symlink"
			view = append(view, e)
			return nil
		case d.IsDir():
			return nil
		case !d.Type().IsRegular():
			e.State, e.Reason = "refused", "not-a-regular-file"
			view = append(view, e)
			return nil
		}
		raw, rerr := os.ReadFile(p)
		if rerr != nil {
			e.State = "unreadable"
		} else {
			sum := sha256.Sum256(raw)
			e.State, e.Sha256 = "read", hex.EncodeToString(sum[:])
		}
		view = append(view, e)
		return nil
	})
	if err != nil {
		return nil, err
	}
	if trustConfig != "" {
		e := entry{Path: filepath.Base(trustConfig), Role: "trust-config"}
		raw, rerr := os.ReadFile(trustConfig)
		if rerr != nil {
			e.State = "unreadable"
		} else {
			sum := sha256.Sum256(raw)
			e.State, e.Sha256 = "read", hex.EncodeToString(sum[:])
		}
		view = append(view, e)
	}
	sort.Slice(view, func(i, j int) bool { return view[i].Path < view[j].Path })
	seen := map[string]bool{}
	for _, e := range view {
		if seen[e.Path] {
			return nil, fmt.Errorf("duplicate path in the observation: %q", e.Path)
		}
		seen[e.Path] = true
	}
	return view, nil
}

func inputRoot(view []entry) string {
	sum := sha256.Sum256([]byte(domain + jcsView(view)))
	return hex.EncodeToString(sum[:])
}

func main() {
	args := os.Args[1:]
	if len(args) < 1 {
		fmt.Fprintln(os.Stderr, "usage: input_manifest <store-dir> [--trust-config PATH] [--root]")
		os.Exit(2)
	}
	trust, rootOnly := "", false
	for i, a := range args {
		if a == "--trust-config" && i+1 < len(args) {
			trust = args[i+1]
		}
		if a == "--root" {
			rootOnly = true
		}
	}
	view, err := seal(args[0], trust)
	if err != nil {
		fmt.Fprintln(os.Stderr, "REFUSED:", err)
		os.Exit(1)
	}
	if rootOnly {
		fmt.Println(inputRoot(view))
		return
	}
	os.Stdout.WriteString(jcsView(view))
}
